#  <span style="color:darkblue; font-size:26px;">**Project "Secrets of the Darkforest"**</span> 
_______________________________________________________________________________________________________________________________

### <span style="color:blue; font-size:18px;"> Project Goals and Objectives</span> 

Familiarization with the database, verification of data accuracy, and preprocessing of data to obtain correct analysis results.

### <span style="color:blue; font-size:18px;"> Data Description </span> 

- **Name** - name of the game  
- **Platform** - name of the platform  
- **Year of Release** - year the game was released  
- **Genre** - game genre  
- **NA sales** - sales in North America (in millions of copies)  
- **EU sales** - sales in Europe (in millions of copies)  
- **JP sales** - sales in Japan (in millions of copies)  
- **Other sales** - sales in other regions (in millions of copies)  
- **Critic Score** - critics’ rating (from 0 to 100)  
- **User Score** - users’ rating (from 0 to 10)  
- **Rating** - ESRB rating (Entertainment Software Rating Board). This organization assigns age-appropriate ratings to video games


### <span style="color:blue; font-size:18px;"> Project Outline </span> 

1) **Data Loading and Familiarization** – loading the dataset and getting acquainted with the data
2) **Data Cleaning and Preprocessing** – this step includes 4 stages, covering the following tasks:
 * Convert column names to snake_case
 * Check and correct data types, replacing invalid values with nulls
 * Handle missing values, check for duplicates in categorical data, and normalize text
 * Remove explicit duplicates and prepare the dataset for analysis

3) **Data Filtering**
4) **Data Categorization** – categorizing data according to the classification in the assignment, identifying the top 7 platforms by number of games, and providing conclusions based on the conducted analysis

### <span style="color:darkviolet; font-size:18px;"> 1. Data Loading and Familiarization </span> 

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('https://code.s3.yandex.net/datasets/new_games.csv')

In [3]:
display(df)

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...
16951,Samurai Warriors: Sanada Maru,PS3,2016.0,Action,0.00,0.0,0.01,0.00,NaN,NaN,NaN
16952,LMA Manager 2007,X360,2006.0,Sports,0.00,0.01,0.0,0.00,NaN,NaN,NaN
16953,Haitaka no Psychedelica,PSV,2016.0,Adventure,0.00,0.0,0.01,0.00,NaN,NaN,NaN
16954,Spirits & Spells,GBA,2003.0,Platform,0.01,0.0,0.0,0.00,NaN,NaN,NaN


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


<span style="color:darkviolet; font-size:18px;">Key Characteristics of the Dataset </span>  

The dataset contains **16,956 rows and 11 columns**.

Missing values are present in **6 columns**:  
* **Name** – 2 missing values  
* **Year of Release** – 275 missing values  
* **Genre** – 2 missing values  
* **Critic Score** – 8,714 missing values (almost half of the dataset)  
* **User Score** – 6,804 missing values  
* **Rating** – 6,871 missing values

For further analysis, these missing values will need to be either handled or taken into account depending on the specific task.

**Data Types:**  
The float64 data type for 4 columns (Year of Release, NA Sales, Other Sales, Critic Score) is appropriate for numerical data.  
The object data type is not correct for all of the remaining 7 columns. In particular: EU Sales and JP Sales should be converted to numeric format.
User Score should also be converted to a numeric type.

**Column Names:**  
The column names reflect their content; however, spaces and capital letters make them inconvenient for coding. Therefore, the column names should be converted to snake_case format.

**Other Dataset Features:**  
* A large number of missing values in Critic Score, User Score, and Rating. It can be assumed that older games may not have this information available.
* Sales columns currently have mixed data types (float and object) and should be converted to a consistent numeric format.  
* There may be duplicates based on game name and platform, which should be checked during preprocessing.
___________________________________________________________________________________________________________________________________

### <span style="color:darkviolet; font-size:18px;"> 2. Data Validation and Preprocessing </span> 

<span style="color:darkviolet; font-size:16px;">2.1. Column Names of the DataFrame </span>  

In [5]:
display("Column Names:")
display(df.columns)

'Column Names:'

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

In [6]:
df.columns = df.columns.str.strip()
df.columns = df.columns.str.lower()
df.columns = df.columns.str.replace(' ', '_')

In [7]:
display("Column Names after converting to snake_case:")
display(df.columns)

'Column Names after converting to snake_case:'

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')

<span style="color:darkviolet; font-size:16px;">2.2. Data Types </span>  

In [8]:
(df.dtypes)

name                object
platform            object
year_of_release    float64
genre               object
na_sales           float64
eu_sales            object
jp_sales            object
other_sales        float64
critic_score       float64
user_score          object
rating              object
dtype: object

<span style="color:darkviolet; font-size:18px;">Interim Conclusion </span>  

The columns in the dataset have the following data types:
1) **Numeric columns:** year_of_release, na_sales, other_sales, critic_score
2) **Categorical columns:** name, platform, genre, eu_sales, jp_sales, user_score, rating

The columns **eu_sales, jp_sales, and user_score,**  have incorrect data type. This issue may be caused by the presence of string values (e.g., "NaN"), which prevent proper conversion to a numeric data type.


In [9]:
# The list of numeric columns
num_cols = ['year_of_release', 'na_sales', 'eu_sales', 'jp_sales', 'other_sales', 'critic_score', 'user_score']

# Convert these columns to float, replacing any invalid string values with NaN.
for col in num_cols:
    df[col] = pd.to_numeric(df[col], errors='coerce')

In [10]:
display(df.dtypes)

name                object
platform            object
year_of_release    float64
genre               object
na_sales           float64
eu_sales           float64
jp_sales           float64
other_sales        float64
critic_score       float64
user_score         float64
rating              object
dtype: object

<span style="color:darkviolet; font-size:16px;">2.3. Missing Values in the Dataset </span>  

In [11]:
# Count the Number
missing_abs = df.isna().sum()
missing_rel = (df.isna().mean() * 100).round(2)

In [12]:
# Combine the Results
missing_data = pd.DataFrame({ 
    'Missing Values (units)': missing_abs,
    'Proportion of Missing Values (%)': missing_rel
}).sort_values(by='Proportion of Missing Values (%)', ascending=False)

In [13]:
display(missing_data)

,Missing Values (units),Proportion of Missing Values (%)
user_score,9268,54.66
critic_score,8714,51.39
rating,6871,40.52
year_of_release,275,1.62
eu_sales,6,0.04
jp_sales,4,0.02
name,2,0.01
genre,2,0.01
platform,0,0.00
na_sales,0,0.00


<span style="color:darkviolet; font-size:18px;">Interim Conclusion </span>  
The largest number of missing values is observed in the following columns:
* **user_score** – 54.66%, **critic_score** – 51.39%  and **rating** – 40.52%
This indicates that critic scores, user scores, and ESRB ratings are missing for more than half of the games.

Smaller amounts of missing data are found in:  
* **year_of_release** – 275 rows (1.62%), **name** and **genre** – 2 missing values each (0.01%), which is minimal but important, as these columns are key for identifying games.
* **eu_sales** and **jp_sales** – 10 rows in total (likely due to invalid string values, e.g., 'NaN').

<span style="color:darkviolet; font-size:18px;">Possible Causes of Missing Values:</span>  

**Historical data:** Older games may not have user or critic scores, especially if released before online rating platforms existed. ESRB ratings were also not assigned to games released outside the US or before 1994.  
**Data collection errors:** When combining data from different sources, some scores may have been missing or recorded as strings like 'tbd' or 'NaN'.   
**Technical issues:** In the eu_sales and jp_sales columns, invalid string formats were converted to NaN during type conversion (errors='coerce').  

<span style="color:darkviolet; font-size:18px;">Handling Missing Values:</span>  

**Key information columns:**  
* **name** – 2, **genre** – 2, **year_of_release** – 275 → remove these rows, as analysis cannot proceed without this information.

**Score columns:**
* **critic_score** – 8,714, **user_score** – 9,268 (more than half of the data) → leave as missing, since imputing such a large volume could distort analysis results.

**Rating column:**
* **rating** – 6,871 → leave missing, as ESRB ratings were not always assigned.

**Sales columns:**
* **eu_sales** – 6, **jp_sales** – 4 → replace missing values with the average sales for the corresponding platform and release year.

In [14]:
df = df.dropna(subset=['name', 'genre', 'year_of_release'])

In [15]:
display(df[['platform', 'year_of_release', 'eu_sales', 'jp_sales']].head(10))

,platform,year_of_release,eu_sales,jp_sales
0,Wii,2006.0,28.96,3.77
1,NES,1985.0,3.58,6.81
2,Wii,2008.0,12.76,3.79
3,Wii,2009.0,10.93,3.28
4,GB,1996.0,8.89,10.22
5,GB,1989.0,2.26,4.22
6,DS,2006.0,9.14,6.50
7,Wii,2006.0,9.18,2.93
8,Wii,2009.0,6.94,4.70
9,NES,1984.0,0.63,0.28


In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16679 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16679 non-null  object 
 1   platform         16679 non-null  object 
 2   year_of_release  16679 non-null  float64
 3   genre            16679 non-null  object 
 4   na_sales         16679 non-null  float64
 5   eu_sales         16673 non-null  float64
 6   jp_sales         16675 non-null  float64
 7   other_sales      16679 non-null  float64
 8   critic_score     8085 non-null   float64
 9   user_score       7558 non-null   float64
 10  rating           9901 non-null   object 
dtypes: float64(7), object(4)
memory usage: 1.5+ MB


<span style="color:darkviolet; font-size:16px;">2.4. Explicit and Implicit Duplicates in the Data </span>  

In [17]:
display("Genre:")
display(df['genre'].unique())

display("Platform:")
display(df['platform'].unique())

display("Rating:")
display(df['rating'].unique())

'Genre:'

array(['Sports', 'Platform', 'Racing', 'Role-Playing', 'Puzzle', 'Misc',
       'Shooter', 'Simulation', 'Action', 'Fighting', 'Adventure',
       'Strategy', 'MISC', 'ROLE-PLAYING', 'RACING', 'ACTION', 'SHOOTER',
       'FIGHTING', 'SPORTS', 'PLATFORM', 'ADVENTURE', 'SIMULATION',
       'PUZZLE', 'STRATEGY'], dtype=object)

'Platform:'

array(['Wii', 'NES', 'GB', 'DS', 'X360', 'PS3', 'PS2', 'SNES', 'GBA',
       'PS4', '3DS', 'N64', 'PS', 'XB', 'PC', '2600', 'PSP', 'XOne',
       'WiiU', 'GC', 'GEN', 'DC', 'PSV', 'SAT', 'SCD', 'WS', 'NG', 'TG16',
       '3DO', 'GG', 'PCFX'], dtype=object)

'Rating:'

array(['E', nan, 'M', 'T', 'E10+', 'K-A', 'AO', 'EC', 'RP'], dtype=object)

In [18]:
# Create a Copy of the DataFrame
df = df.copy()
# Normalization of Genres and Platforms
df['genre'] = df['genre'].str.lower().str.strip()
df['platform'] = df['platform'].str.upper().str.strip()

display("Genre:")
display(df['genre'].unique())

display("Platform:")
display(df['platform'].unique())

'Genre:'

array(['sports', 'platform', 'racing', 'role-playing', 'puzzle', 'misc',
       'shooter', 'simulation', 'action', 'fighting', 'adventure',
       'strategy'], dtype=object)

'Platform:'

array(['WII', 'NES', 'GB', 'DS', 'X360', 'PS3', 'PS2', 'SNES', 'GBA',
       'PS4', '3DS', 'N64', 'PS', 'XB', 'PC', '2600', 'PSP', 'XONE',
       'WIIU', 'GC', 'GEN', 'DC', 'PSV', 'SAT', 'SCD', 'WS', 'NG', 'TG16',
       '3DO', 'GG', 'PCFX'], dtype=object)

In [19]:
# Check for Explicit Duplicates
duplicates = df.duplicated()

display("Number of explicit duplicates:", duplicates.sum())

'Number of explicit duplicates:'

235

In [20]:
# Remove Duplicates
df = df.drop_duplicates()

display("New DataFrame size after removing duplicates:", df.shape)

'New DataFrame size after removing duplicates:'

(16444, 11)

<span style="color:darkviolet; font-size:18px;">Interim Conclusion </span>  
During the data inspection, 235 explicit duplicates were identified. To clean the dataset, all duplicates were removed, reducing the DataFrame size from 16,679 rows to 16,444 rows.  

Additional data normalization steps were performed:  
* Genres were converted to lowercase and stripped of extra spaces  
* Platforms were converted to uppercase and stripped of extra spaces
* ESRB Rating were converted to uppercase, with missing values replaced by 'UNKNOWN'.

In [21]:
df = pd.read_csv('https://code.s3.yandex.net/datasets/new_games.csv')
df.columns = df.columns.str.strip()
df.columns = df.columns.str.lower()
df.columns = df.columns.str.replace(' ', '_')

# Original Number of Rows
original_rows = len(df)

# Data Cleaning
num_after_missingcleaned = 16679 
deleted_missing = original_rows - num_after_missingcleaned 
deleted_missing_percent = (deleted_missing / original_rows) * 100
num_after_duplicatecleaned = 16444 
deleted_duplicates = num_after_missingcleaned - num_after_duplicatecleaned 
deleted_duplicates_percent = (deleted_duplicates / original_rows) * 100  


# Final Number of Rows
final_rows = num_after_duplicatecleaned
total_deleted = original_rows - final_rows
total_deleted_percent = (total_deleted / original_rows) * 100

In [22]:
display("=== Data Cleaning Report ===")
display(f"Original Number of Rows: {original_rows}")
display(f"Number of Rows Removed Due to Missing Values: {deleted_missing} ({deleted_missing_percent:.2f}%)")
display(f"Number of Rows Removed Due to Duplicates: {deleted_duplicates} ({deleted_duplicates_percent:.2f}%)")
display(f"Final Number of Rows: {final_rows}")
display(f"Total Number of Rows Removed: {total_deleted} ({total_deleted_percent:.2f}%)")

'=== Data Cleaning Report ==='

'Original Number of Rows: 16956'

'Number of Rows Removed Due to Missing Values: 277 (1.63%)'

'Number of Rows Removed Due to Duplicates: 235 (1.39%)'

'Final Number of Rows: 16444'

'Total Number of Rows Removed: 512 (3.02%)'

<span style="color:darkviolet; font-size:18px;">Interim Conclusion </span>  
* Original Dataset Size: 16,956 rows  
* Rows Removed Due to Missing Values: 277 (~1.63% of the original dataset) – eliminated incomplete or incorrect entries that could distort analysis.  
* Rows Removed Due to Duplicates: 235 (~1.39% of the original dataset) – ensured uniqueness of records, important for accurate statistics and visualization.  
* Final Dataset Size: 16,444 rows
* Total Rows Removed: 512 (~3.02% of the original dataset)
________________________________________________________________________________________________________________________________________________________

### <span style="color:darkviolet; font-size:18px;"> 3. Data Filtering </span> 

In [23]:
df_actual = df[(df['year_of_release'] >= 2000) & (df['year_of_release'] <= 2013)].copy()

In [24]:
display(df_actual.head())
display(f"Number of Records:{len(df_actual)}")

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
6,New Super Mario Bros.,DS,2006.0,Platform,11.28,9.14,6.5,2.88,89.0,8.5,E
7,Wii Play,Wii,2006.0,Misc,13.96,9.18,2.93,2.84,58.0,6.6,E


'Number of Records:12980'

_______________________________________________________________________________________________________________________________________________________

### <span style="color:darkviolet; font-size:18px;"> 4. Data Categorization </span> 

In [25]:
df_actual['user_score'] = pd.to_numeric(df_actual['user_score'], errors='coerce')
# Categorization of Scores
def categorize_rating(score):
    if 8 <= score <= 10:
        return "High Score"
    elif 3 <= score < 8:
        return "Medium Score"
    elif 0 <= score < 3:
        return "Low Score"
    else:
        return "Invalid Score"  

In [26]:
# Apply the Categorization Function to df_actual
df_actual['user_score_category'] = df_actual['user_score'].apply(categorize_rating)

In [27]:
# Checking the Result
display("Examples of Games by Score Category user_score (2000–2013)")
display(df_actual[['name', 'year_of_release', 'user_score', 'user_score_category']].head())

display("Distribution of Games by Score Categories user_score (2000–2013)")
display(df_actual['user_score_category'].value_counts())

'Examples of Games by Score Category user_score (2000–2013)'

,name,year_of_release,user_score,user_score_category
0,Wii Sports,2006.0,8.0,High Score
2,Mario Kart Wii,2008.0,8.3,High Score
3,Wii Sports Resort,2009.0,8.0,High Score
6,New Super Mario Bros.,2006.0,8.5,High Score
7,Wii Play,2006.0,6.6,Medium Score


'Distribution of Games by Score Categories user_score (2000–2013)'

user_score_category
Invalid Score    6408
Medium Score     4148
High Score       2307
Low Score         117
Name: count, dtype: int64

In [28]:
# Convert critic_score to Numeric, invalid data will going to NaN
df_actual['critic_score'] = pd.to_numeric(df_actual['critic_score'], errors='coerce')
# Categorization Function for Critic Scores
def categorize_critic(score):
    if pd.isna(score):
        return "No Score"
    elif 80 <= score <= 100:
        return "High Score"
    elif 30 <= score < 80:
        return "Medium Score"
    elif 0 <= score < 30:
        return "Low Score"
    else:
        return "No Score"

In [29]:
# Apply the Function
df_actual['critic_score_category'] = df_actual['critic_score'].apply(categorize_critic)

In [30]:
# Checking the Result
display("Examples of Games by Score Category critic_score (2000–2013)")
display(df_actual[['name', 'year_of_release', 'critic_score', 'critic_score_category']].head())

display("Distribution of Games by Score Categories critic_score (2000–2013)")
display(df_actual['critic_score_category'].value_counts())

'Examples of Games by Score Category critic_score (2000–2013)'

,name,year_of_release,critic_score,critic_score_category
0,Wii Sports,2006.0,76.0,Medium Score
2,Mario Kart Wii,2008.0,82.0,High Score
3,Wii Sports Resort,2009.0,80.0,High Score
6,New Super Mario Bros.,2006.0,89.0,High Score
7,Wii Play,2006.0,58.0,Medium Score


'Distribution of Games by Score Categories critic_score (2000–2013)'

critic_score_category
No Score        5713
Medium Score    5500
High Score      1712
Low Score         55
Name: count, dtype: int64

In [32]:
# Convert to Numeric Type
df_actual['user_score'] = pd.to_numeric(df_actual['user_score'], errors='coerce')
df_actual['critic_score_category'] = df_actual['critic_score'].apply(categorize_critic)

# Categorization Function for User Scores
def categorize_user(score):
    if pd.isna(score):
        return "No Score"
    elif 8 <= score <= 10:
        return "High Score"
    elif 3 <= score < 8:
        return "Medium Score"
    elif 0 <= score < 3:
        return "Low Score"
    else:
        return "No Score"

# Categorization Function for Critic Scores
def categorize_critic(score):
    if pd.isna(score):
        return "No Score"
    elif 80 <= score <= 100:
        return "High Score"
    elif 30 <= score < 80:
        return "Medium Score"
    elif 0 <= score < 30:
        return "Low Score"
    else:
        return "No Score"

# Create New Columns for Score Categories
df_actual['user_rating_category'] = df_actual['user_score'].apply(categorize_user)
df_actual['critic_rating_category'] =df_actual['critic_score'].apply(categorize_critic)

In [33]:
user_counts = df_actual.groupby('user_score_category').size()
critic_counts = df_actual.groupby('critic_score_category').size()

In [34]:
display("Number of Games by User Score Categories:")
display(user_counts)
display("Number of Games by Critic Score Categories:")
display(critic_counts)

'Number of Games by User Score Categories:'

user_score_category
High Score       2307
Invalid Score    6408
Low Score         117
Medium Score     4148
dtype: int64

'Number of Games by Critic Score Categories:'

critic_score_category
High Score      1712
Low Score         55
Medium Score    5500
No Score        5713
dtype: int64

In [35]:
# Group by Platform and Count Games
platform_counts = df_actual['platform'].value_counts()

In [36]:
# Sort by Descending and Select Top 7 Platforms
top_7_platforms = platform_counts.head(7)

In [37]:
display("Top 7 Platforms by Number of Games (2000–2013)")
display(top_7_platforms)

'Top 7 Platforms by Number of Games (2000–2013)'

platform
PS2     2154
DS      2146
Wii     1294
PSP     1199
X360    1138
PS3     1107
GBA      826
Name: count, dtype: int64

________________________________________________________________________________________________________________________________________________________

### <span style="color:darkviolet; font-size:18px;"> Final Conclusion </span>   

The original dataset was reduced by 512 rows due to the removal of duplicates and missing values.  

A time-based filter was applied for the period from 2000 to 2013 inclusive, and the resulting subset was saved in a separate DataFrame called df_actual to enable analysis of game sales history in the early 21st century.

The final step involved data categorization, specifically identifying the TOP-7 platforms by the number of games. Among these, three positions are occupied by PlayStation platforms, indicating their popularity among players during the analyzed period.  

The final step was to categorize the data, identifying the top seven platforms by number of games. PS platforms occupy three positions in the provided list, demonstrating their popularity among gamers.